dd-java-agent/instrumentation/jms/src/main/java/datadog/trace/instrumentation/jms/MDBMessageConsumerInstrumentation.java

```diff
@@ -0,0 +1,97 @@
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+import com.google.auto.service.AutoService;
+import datadog.trace.agent.tooling.Instrumenter;
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import datadog.trace.bootstrap.instrumentation.api.AgentSpan;
+import javax.jms.Destination;
+import javax.jms.JMSException;
+import javax.jms.Message;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+@AutoService(Instrumenter.class)
+public final class MDBMessageConsumerInstrumentation extends Instrumenter.Tracing
+    implements Instrumenter.ForTypeHierarchy {
+
+  public MDBMessageConsumerInstrumentation() {
+    super("javax-mdb");
+  }
+
+  @Override
+  public String[] helperClassNames() {
+    return new String[] {
+      packageName + ".JMSDecorator",
+      packageName + ".MessageExtractAdapter",
+      packageName + ".MessageExtractAdapter$1"
+    };
+  }
+
+  @Override
+  public String hierarchyMarkerType() {
+    return "javax.jms.MessageListener";
+  }
+
+  @Override
+  public ElementMatcher<TypeDescription> hierarchyMatcher() {
+    return implementsInterface(named(hierarchyMarkerType()))
+        .and(
+            hasSuperType(declaresAnnotation(named("javax.ejb.MessageDriven")))
+                .or(implementsInterface(named("javax.ejb.MessageDrivenBean"))));
+  }
+
+  @Override
+  public void adviceTransformations(AdviceTransformation transformation) {
+    transformation.applyAdvice(
+        isMethod()
+            .and(isPublic())
+            .and(named("onMessage"))
+            .and(takesArguments(1))
+            .and(takesArgument(0, (named("javax.jms.Message")))),
+        getClass().getName() + "$MDBAdvice");
+  }
+
+  public static class MDBAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static AgentScope methodEnter(@Advice.Argument(0) final Message message) {
+      AgentSpan.Context propagatedContext = propagate().extract(message, GETTER);
+      AgentSpan span = startSpan(JMS_CONSUME, propagatedContext);
+      CONSUMER_DECORATE.afterStart(span);
+      CharSequence consumerResourceName;
+      try {
+        Destination destination = message.getJMSDestination();
+        boolean isQueue = CONSUMER_DECORATE.isQueue(destination);
+        String destinationName = CONSUMER_DECORATE.getDestinationName(destination);
+        consumerResourceName = CONSUMER_DECORATE.toResourceName(destinationName, isQueue);
+      } catch (JMSException e) {
+        logJMSException(e);
+        consumerResourceName = "unknown JMS destination";
+      }
+      CONSUMER_DECORATE.onConsume(span, message, consumerResourceName);
+      return activateSpan(span);
+    }
+
+    @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+    public static void methodExit(@Advice.Enter AgentScope scope) {
+      scope.close();
+      scope.span().finish();
```

Comment:
```suggestion
      if (null != scope) {
        scope.close();
        scope.span().finish();
      }
```

just in case something happens inside the enter advice which means the span isn't activated